gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/GValue.java

```diff
@@ -0,0 +1,446 @@
+     * can directly check the type, otherwise it will test the given object's class itself using the mapping on the
+     * {@link GType}.
+     */
+    public static boolean instanceOf(final Object o, final GType type) {
+        if (null == o)
+            return false;
+        else if (o instanceof GValue)
+            return ((GValue) o).getType() == type;
+        else
+            return type.getJavaType().isAssignableFrom(o.getClass());
+    }
+
+    /**
+     * Returns {@code true} if the object is a collection or a {@link GValue} that contains a {@link Collection}.
+     */
+    public static boolean instanceOfCollection(final Object o) {
+        return o instanceof Collection || (o instanceof GValue && ((GValue) o).getType().isCollection());
+    }
+
+    /**
+     * Returns {@code true} if the object is an element or a {@link GValue} that contains an {@link Element}.
+     */
+    public static boolean instanceOfElement(final Object o) {
+        return o instanceof Element || (o instanceof GValue && ((GValue) o).getType().isElement());
+    }
+
+    /**
+     * Returns {@code true} if the object is a number or a {@link GValue} that contains a number.
+     */
+    public static boolean instanceOfNumber(final Object o) {
+        return o instanceof Number || (o instanceof GValue && ((GValue) o).getType().isNumeric());
+    }
+
+    /**
+     * The elements in object array argument are examined to see if they are {@link GValue} objects. If they are, they
+     * are preserved as is. If they are not then they are wrapped in a {@link GValue} object.
+     */
+    public static <T> GValue<T>[] ensureGValues(final Object[] args) {
+        return Stream.of(args).map(GValue::of).toArray(GValue[]::new);
+    }
+
+    /**
+     * Converts {@link GValue} objects argument array to their values to prevent them from leaking to the Graph API.
+     * Providers extending from this step should use this method to get actual values to prevent any {@link GValue}
+     * objects from leaking to the Graph API.
+     */
+    public static Object[] resolveToValues(final GValue<?>[] gvalues) {
+        final Object[] values = new Object[gvalues.length];
+        for (int i = 0; i < gvalues.length; i++) {
+            values[i] = gvalues[i].get();
+        }
+        return values;
+    }
+
+    /**
+     * Takes an argument that is either a {@code GValue} or an object and if the former, returns the child object and
+     * if the latter returns the object itself.
+     */
+    public static Object valueOf(final Object either) {
+        if (either instanceof GValue)
+            return ((GValue<?>) either).get();
+        else
+            return either;
+    }
+
+    /**
+     * Takes an argument that is either a {@code GValue} or an object and returns its {@code Number} form, otherwise
+     * throws an exception.
+     */
+    public static Number numberOf(final Object either) {
```

Comment:
inconsistency with `GType.isNumeric()`
`Float` GValue is number, but not `isNumeric()`

Reply:
I've added additional tests to ensure that `numberOf()` is handling all numeric GTypes including Float.